### PR TITLE
Added problems for C12

### PR DESCRIPTION
**Pull request by Grizzlly**

### CODE REVIEW

File: Curs12/C/01-macros.c [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ *		Exemple de utilizare
+ *
+ */
+
+#include <stdio.h>
+
+#define BAD_ADD(a, b) a + b
+#define GOOD_ADD(a, b) ((a) + (b))
+
+#define BAD_MUL(a, b) a *b
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:13]:
```suggestion
#define BAD_MUL(a, b) a * b
```

File: Curs12/C/01-macros.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+#define ADD_4(a, ...) ((a) + ADD_3(__VA_ARGS__))
+#define ADD_5(a, ...) ((a) + ADD_4(__VA_ARGS__))
+
+#define EXPAND(...) __VA_ARGS__
+
+int main()
+{
+	int a, b;
+
+	scanf("%d %d", &a, &b);
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:28]:
```suggestion
	scanf("%d%d", &a, &b);
```

File: Curs12/C/01-macros.c [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/* Programarea Calculatoarelor CA, 2022-2023
+ *
+ *	Macro definitii pentru suma si produs
+ *		Exemple de utilizare
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:4]:
```suggestion
```

File: Curs12/C/01-macros.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+#include <stdio.h>
+
+#define BAD_ADD(a, b) a + b
+#define GOOD_ADD(a, b) ((a) + (b))
+
+#define BAD_MUL(a, b) a *b
+#define GOOD_MUL(a, b) ((a) * (b))
+
+// Happy Halloween!
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:16]: why do we have this comment?

File: Curs12/C/01-macros.c [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	 *
+	 *	https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms
+	 *	https://github.com/pfultz2/Cloak/wiki/Is-the-C-preprocessor-Turing-complete%3F
+	 */
+
+	return 0;
+}
+
+/*
+ *	compile: gcc 01-macros.c -o macros
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:58]: add an example of what is the code if you stop after the preprocessing phase (gcc -E)

File: Curs12/C/03-params_sum.c [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ *	Suma elementelor parametrilor liniei de comanda
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+int parseInt(char *str)
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:11]:
```suggestion
int parseInt(const char *str)
```

Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:11]: I think we may use atoi?

File: Curs12/C/03-params_sum.c [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+			return 0; // NaN
+
+	return number;
+}
+
+int main(int argc, char **argv)
+{
+	int s = 0;
+
+	for (int i = 0; i < argc; i++)
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:29]: why do we start here from 0?

File: Curs12/C/01-macros.c [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		   GOOD_MUL(a + b, a + b));
+
+	// How many variables can I have?
+	printf("1 + 2 + 3 + 4 = %d can someone explain?\n",
+		   EXPAND(EXPAND(EXPAND(EXPAND(ADD_4(1, 2, 3, 4))))));
+
+	/*
+	 *	Conclusion:
+	 *
+	 *		Take the pre-processor for what it is: FIND & REPLACE. No more and no less
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:48]:
```suggestion
	 *	Take the pre-processor for what it is: FIND & REPLACE. No more and no less
```

File: Curs12/C/03-params_sum.c [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	for (int i = 0; i < strsize; i++)
+		if ((str[i] <= '9') && (str[i] >= '0'))
+			number = number * 10 + (int)(str[i] - '0');
+		else
+			return 0; // NaN
+
+	return number;
+}
+
+int main(int argc, char **argv)
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:25]:
```suggestion
int main(int argc, const char **argv)
```

File: Curs12/C/01-macros.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+#define GOOD_ADD(a, b) ((a) + (b))
+
+#define BAD_MUL(a, b) a *b
+#define GOOD_MUL(a, b) ((a) * (b))
+
+// Happy Halloween!
+#define ADD_2(a, b) ((a) + (b))
+#define ADD_3(a, ...) ((a) + ADD_2(__VA_ARGS__))
+#define ADD_4(a, ...) ((a) + ADD_3(__VA_ARGS__))
+#define ADD_5(a, ...) ((a) + ADD_4(__VA_ARGS__))
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:20]: do you use `ADD_5` somewhere?

File: Curs12/C/01-macros.c [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+#define EXPAND(...) __VA_ARGS__
+
+int main()
+{
+	int a, b;
+
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:27]:
```suggestion
```

File: Curs12/C/01-macros.c [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+// Happy Halloween!
+#define ADD_2(a, b) ((a) + (b))
+#define ADD_3(a, ...) ((a) + ADD_2(__VA_ARGS__))
+#define ADD_4(a, ...) ((a) + ADD_3(__VA_ARGS__))
+#define ADD_5(a, ...) ((a) + ADD_4(__VA_ARGS__))
+
+#define EXPAND(...) __VA_ARGS__
+
+int main()
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:24]:
```suggestion
int main(void)
```

**Review by RaduNichita — CHANGES_REQUESTED** [review]
Thanks for the contribution, @Grizzlly! 💯  

Please see the comments

File: Curs12/C/03-params_sum.c [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ *	Suma elementelor parametrilor liniei de comanda
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+int main(int argc, const char **argv)
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:11]:
```suggestion
int main(const int argc, const char **argv)
```

File: Curs12/C/01-macros.c [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+ *		BAD_MUL:
+ *				(10 + 12) * (10 + 12) = 142
+ *
+ *		GOOD_MUL:
+ *				(10 + 12) * (10 + 12) = 484
+ *
+ *		1 + 2 + 3 + 4 = 10 can someone explain?
+ *
+ * 	preprocessing phase: gcc -E 01-macros.c
+ * 		int main(void)
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:75]: You should add some ... before to be clear that there is some other code that is generated by the compiler.

File: Curs12/C/01-macros.c [hunk 15]
```diff
@@ -0,0 +1,95 @@
+/* Programarea Calculatoarelor CA, 2022-2023
+ *
+ *	Macro definitii pentru suma si produs
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:3]:
```suggestion
 *	Diferite macro definitii pentru suma si produs
```

File: Curs12/C/03-params_sum.c [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+{
+	int s = 0;
+
+	for (int i = 1; i < argc; i++)
+		s += atoi(argv[i]);
+
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:17]:
```suggestion
```

**Review by RaduNichita — CHANGES_REQUESTED** [review]
@Grizzlly, please see my comments.

File: Curs12/C/01-macros.c [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+	// How many variables can I have?
+	printf("1 + 2 + 3 + 4 = %d can someone explain?\n",
+		   EXPAND(EXPAND(EXPAND(EXPAND(ADD_4(1, 2, 3, 4))))));
+
+	/*
+	 *	Conclusion:
+	 *
+	 *	Take the pre-processor for what it is: FIND & REPLACE. No more and no less
+	 *
+	 *	https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:46]:
```suggestion
     *  Further reading:
	 * 
	 *	https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms
```

File: Curs12/C/01-macros.c [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+	 *	https://github.com/pfultz2/Cloak/wiki/Is-the-C-preprocessor-Turing-complete%3F
+	 */
+
+	return 0;
```
Comment (RaduNichita) [inline Curs12/C/01-macros.c:50]: put the links after return 0.

**Review by RaduNichita — CHANGES_REQUESTED** [review]
I will approve after you fix the problems

File: Curs12/C/03-params_sum.c [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+	for (int i = 1; i < argc; i++)
+		s += atoi(argv[i]);
+	printf("%d\n", s);
+
+	return 0;
+}
+
+/*
+ *	compile: gcc 03-params_sum.c -o params_sum
+ *	run: ./params_sum 1 2 3 4 5
```
Comment (alexandrubuzea) [inline Curs12/C/03-params_sum.c:24]: Maybe check if the CLI params (1 <= i <= argc - 1) are numbers ? (aka defensive programming before using atoi())

**Comment by alexandrubuzea** [comment]
Looks good, good job @Grizzlly !

File: Curs12/C/03-params_sum.c (hunk 19 above)
Comment (RaduNichita) [reply Curs12/C/03-params_sum.c:24]: yes, as this would be an invalid input

**Review by RaduNichita — CHANGES_REQUESTED** [review]
We can merge after this check, @Grizzlly

File: Curs12/C/03-params_sum.c [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+	return true;
+}
+
+int main(const int argc, const char **argv)
+{
+	int s = 0;
+
+	for (int i = 1; i < argc; i++) {
+		if (!isnumber(argv[i])) {
+			puts("Invalid input");
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:28]: fprintf(stderr, "Invalid input");

File: Curs12/C/03-params_sum.c [hunk 21]
```diff
@@ -0,0 +1,46 @@
+/* Programarea Calculatoarelor CA, 2022-2023
+ *
+ *	Suma elementelor parametrilor liniei de comanda
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:10]:
```suggestion
```

File: Curs12/C/03-params_sum.c [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+#include <ctype.h>
+
+bool isnumber(const char * str)
+{
+	for (int i = 0; str[i] != '\0'; i++)
+		if (!isdigit(str[i]))
+			return false;
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:17]:
```suggestion
			return 0;
```

File: Curs12/C/03-params_sum.c [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+#include <ctype.h>
+
+bool isnumber(const char * str)
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:13]:
```suggestion
int isnumber(const char * str)
```

File: Curs12/C/03-params_sum.c [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+{
+	for (int i = 0; str[i] != '\0'; i++)
+		if (!isdigit(str[i]))
+			return false;
+
+	return true;
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:19]:
```suggestion
	return 1;
```

File: Curs12/C/03-params_sum.c [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+int main(const int argc, const char **argv)
```
Comment (RaduNichita) [inline Curs12/C/03-params_sum.c:22]: Sorry, I changed my mind. Could you please have 

```suggestion
int main(int argc, char *argv[])
```

**Review by RaduNichita — CHANGES_REQUESTED** [review]
Last round

File: Curs12/C/03-params_sum.c (hunk 25 above)
Comment (Grizzlly) [reply Curs12/C/03-params_sum.c:22]: why this change?

Comment (dorinelfilip) [reply Curs12/C/03-params_sum.c]: The [C standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2310.pdf), per chapter 5.2.2.1, allows any equivalent of `int main(int argc, char *argv[])`, including `int main(int argc, char **argv)`, but with no reference to the use of `const` or other modifiers. 

Thus, I would recomand either `int argc, char *argv[]` or `int argc, char **argv[]`.

Comment (dorinelfilip) [reply Curs12/C/03-params_sum.c]: The [C standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2310.pdf), per chapter 5.1.2.2.1 (page 11), allows any equivalent of `int main(int argc, char *argv[])`, including `int main(int argc, char **argv)`, but with no reference to the use of `const` or other modifiers. 

Thus, I would recomand either `int argc, char *argv[]` or `int argc, char **argv[]`.

**Review by RaduNichita — APPROVED** [review]
All good, thanks for the effort, @Grizzlly  :1st_place_medal: 

Approved-by: Radu Nichita radunichita99@gmail.com

**Pull request merged**